<robot name="robot_model" xmlns:xacro="http://www.ros.org/wiki/xacro">
    
  <xacro:property name="gaussian_noise_mean" value="$(arg gaussian_noise_mean)" />


<!-- ROBOT LINKS AND JOINT INFORMATION (MASS, INERTIAS, RELATIONSHIP, ETC.) -->
  <link name="dummy">
  </link>
  <joint name="dummy_joint" type="fixed">
    <parent link="dummy"/>
    <child link="base_link"/>
  </joint>
  <link name="base_link">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="13.189" />
      <inertia ixx="0.44216" ixy="-3.7455E-11" ixz="3.9704E-09" iyy="0.87232" iyz="-2.1794E-05" izz="1.2206" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://robot_model/meshes/base_link.STL" />
      </geometry>
      <material name="">
        <color rgba="1 0 0 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://robot_model/meshes/base_link.STL" />
      </geometry>
    </collision>
  </link>
  <link name="Link1">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="3.2337" />
      <inertia ixx="0.071364" ixy="-9.0285E-09" ixz="3.4807E-10" iyy="0.049192" iyz="1.6935E-08" izz="0.089213" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://robot_model/meshes/Link1.STL" />
      </geometry>
      <material name="">
        <color rgba="0 1 0 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://robot_model/meshes/Link1.STL" />
      </geometry>
    </collision>
  </link>
  <joint name="Joint1" type="revolute">
    <origin xyz="0 0 0" rpy="0 0 0" />
    <parent link="base_link" />
    <child link="Link1" />
    <axis xyz="0 0 -1" />
    <limit lower="-3.1416" upper="3.1416" effort="100" velocity="50" />
  </joint>
  <link name="Link2">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="4.7034" />
      <inertia ixx="0.054672" ixy="1.3667E-07" ixz="-0.014278" iyy="0.29069" iyz="5.385E-08" izz="0.25542" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://robot_model/meshes/Link2.STL" />
      </geometry>
      <material name="">
        <color rgba="0 0 1 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://robot_model/meshes/Link2.STL" />
      </geometry>
    </collision>
  </link>
  <joint name="Joint2" type="revolute">
    <origin xyz="0 -0.032 0" rpy="1.5708 -1.5708 0" />
    <parent link="Link1" />
    <child link="Link2" />
    <axis xyz="0 0 1" />
    <limit lower="-3.14159" upper="3.14159" effort="100" velocity="50" />
  </joint>
  <link name="Link3">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="1.398" />
      <inertia ixx="0.010442" ixy="1.0595E-08" ixz="-0.0012119" iyy="0.012215" iyz="-5.6993E-09" izz="0.0035247" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://robot_model/meshes/Link3.STL" />
      </geometry>
      <material name="">
        <color rgba="0.25 0.25 0.25 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://robot_model/meshes/Link3.STL" />
      </geometry>
    </collision>
  </link>
  <joint name="Joint3" type="revolute">
    <origin xyz="0.6 0 0" rpy="0 0 0" />
    <parent link="Link2" />
    <child link="Link3" />
    <axis xyz="0 0 1" />
    <limit lower="-3.14159" upper="3.14159" effort="100" velocity="50" />
  </joint>
  <link name="Link4">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="1.8144" />
      <inertia ixx="0.028059" ixy="4.6E-09" ixz="-1.5764E-09" iyy="0.028173" iyz="-0.0007752" izz="0.0033203" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://robot_model/meshes/Link4.STL" />
      </geometry>
      <material name="">
        <color rgba="0.25 0.25 0.5 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://robot_model/meshes/Link4.STL" />
      </geometry>
    </collision>
  </link>
  <joint name="Joint4" type="revolute">
    <origin xyz="0 0 0.1175" rpy="1.5708 0 1.5708" />
    <parent link="Link3" />
    <child link="Link4" />
    <axis xyz="0 0 1" />
    <limit lower="-3.14159" upper="3.14159" effort="100" velocity="50" />
  </joint>
  <link name="Link5">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="1.1558" />
      <inertia ixx="0.003775" ixy="-1.1827E-09" ixz="-1.2426E-09" iyy="0.0038772" iyz="0.00017257" izz="0.0014527" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://robot_model/meshes/Link5.STL" />
      </geometry>
      <material name="">
        <color rgba="0.5 0.25 0.25 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://robot_model/meshes/Link5.STL" />
      </geometry>
    </collision>
  </link>
  <joint name="Joint5" type="revolute">
    <origin xyz="0 0 0.5" rpy="-1.5708 0 0" />
    <parent link="Link4" />
    <child link="Link5" />
    <axis xyz="0 0 -1" />
    <limit lower="-3.14159" upper="3.14159" effort="100" velocity="50" />
  </joint>
  <link name="Link6">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="1.4646" />
      <inertia ixx="0.0092037" ixy="-3.4431E-20" ixz="1.5284E-19" iyy="0.0093141" iyz="-1.1154E-19" izz="0.0010717" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://robot_model/meshes/Link6_glass.STL" />
      </geometry>
      <material name="">
        <color rgba="0.25 0.5 0.25 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://robot_model/meshes/Link6_glass.STL" />
      </geometry>
    </collision>
  </link>
  <joint name="Joint6" type="revolute">
    <origin xyz="0 -0.3857 -0.1175" rpy="1.5708 0 0" />
    <parent link="Link5" />
    <child link="Link6" />
    <axis xyz="0 0 1" />
    <limit lower="-3.14159" upper="3.14159" effort="100" velocity="50" />
  </joint>
  
  <link name="link_leftFrontWheel">
        <inertial>
            <origin xyz="-5.5511E-17 2.2204E-16 0.002106" rpy="0 0 0" />
            <mass value="1.0489" />
            <inertia ixx="0.006146" ixy="2.1218E-17" ixz="1.1965E-18" iyy="0.006146" iyz="2.407E-19" izz="0.010259" />
        </inertial>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry>
                <mesh filename="package://robot_model/meshes/link_leftFrontWheel.STL" />
            </geometry>
            <material name="">
                <color rgba="0.93 0.93 0 1" />
            </material>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry>
                <mesh filename="package://robot_model/meshes/link_leftFrontWheel.STL" />
            </geometry>
        </collision>
    </link>
    <joint name="joint_leftFrontWheel" type="continuous">
        <origin xyz="0.315 0.3525 -0.32235" rpy="1.5708 0 0" />
        <parent link="base_link" />
        <child link="link_leftFrontWheel" />
        <axis xyz="0 0 -1" />
    </joint>
    <link name="link__rightFrontWheel">
        <inertial>
            <origin xyz="0 -2.2204E-16 -0.002106" rpy="0 0 0" />
            <mass value="1.0489" />
            <inertia ixx="0.006146" ixy="-2.0382E-17" ixz="-7.0505E-19" iyy="0.006146" iyz="1.0064E-18" izz="0.010259" />
        </inertial>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry>
                <mesh filename="package://robot_model/meshes/link__rightFrontWheel.STL" />
            </geometry>
            <material name="">
                <color rgba="0.93 0.93 0 1" />
            </material>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry>
                <mesh filename="package://robot_model/meshes/link__rightFrontWheel.STL" />
            </geometry>
        </collision>
    </link>
    <joint name="joint__rightFrontWheel" type="continuous">
        <origin xyz="0.315 -0.3525 -0.32235" rpy="1.5708 0 0" />
        <parent link="base_link" />
        <child link="link__rightFrontWheel" />
        <axis xyz="0 0 -1" />
    </joint>
    <link name="link__rightRearWheel">
        <inertial>
            <origin xyz="-2.2204E-16 -1.1102E-16 0.002106" rpy="0 0 0" />
            <mass value="1.0489" />
            <inertia ixx="0.006146" ixy="-8.1049E-18" ixz="-3.7604E-19" iyy="0.006146" iyz="-5.8935E-19" izz="0.010259" />
        </inertial>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry>
                <mesh filename="package://robot_model/meshes/link__rightRearWheel.STL" />
            </geometry>
            <material name="">
                <color rgba="0.93 0.93 0 1" />
            </material>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry>
                <mesh filename="package://robot_model/meshes/link__rightRearWheel.STL" />
            </geometry>
        </collision>
    </link>
    <joint name="joint__rightRearWheel" type="continuous">
        <origin xyz="-0.315 -0.3525 -0.32235" rpy="1.5708 0 0" />
        <parent link="base_link" />
        <child link="link__rightRearWheel" />
        <axis xyz="0 0 -1" />
        <limit lower="0" upper="0" effort="100" velocity="100" />
    </joint>
    <link name="link__leftRearWheel">
        <inertial>
            <origin xyz="0 -1.9429E-16 -0.002106" rpy="0 0 0" />
            <mass value="1.0489" />
            <inertia ixx="0.006146" ixy="-2.0871E-17" ixz="-4.7748E-19" iyy="0.006146" iyz="8.2626E-19" izz="0.010259" />
        </inertial>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry>
                <mesh filename="package://robot_model/meshes/link__leftRearWheel.STL" />
            </geometry>
            <material name="">
                <color rgba="0.93 0.93 0 1" />
            </material>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry>
                <mesh filename="package://robot_model/meshes/link__leftRearWheel.STL" />
            </geometry>
        </collision>
    </link>
    <joint name="joint__leftRearWheel" type="continuous">
        <origin xyz="-0.315 0.3525 -0.32235" rpy="1.5708 0 0" /> 
        <parent link="base_link" />
        <child link="link__leftRearWheel" />
        <axis xyz="0 0 -1" />
        <limit lower="0" upper="0" effort="100" velocity="100" />
    </joint>
  
  
 <!-- GAZEBO PARAMETERS FOR EACH LINK (COLLISION, WHEEL FRICTION) -->
  <gazebo reference="base_link">
    <material>Gazebo/Red</material>
    <self_collide>1</self_collide>
  </gazebo>
  
  <gazebo reference="link__leftRearWheel">
    <material>Gazebo/Yellow</material>
    <self_collide>1</self_collide>
    <mu1>0.2</mu1>
    <mu2>1.0</mu2>
  </gazebo>
  
  <gazebo reference="link__rightRearWheel">
    <material>Gazebo/Yellow</material>
    <self_collide>1</self_collide>
    <mu1>0.2</mu1>
    <mu2>1.0</mu2>
  </gazebo>
  
  <gazebo reference="link_leftFrontWheel">
    <material>Gazebo/Yellow</material>
    <self_collide>1</self_collide>
    <mu1>0.2</mu1>
    <mu2>1.0</mu2>
  </gazebo>
  
  <gazebo reference="link__rightFrontWheel">
    <material>Gazebo/Yellow</material>
    <self_collide>1</self_collide>
    <mu1>0.2</mu1>
    <mu2>1.0</mu2>
  </gazebo>
  
  <gazebo reference="Link1">
    <material>Gazebo/Green</material>
    <self_collide>1</self_collide>
  </gazebo>
  
  <gazebo reference="Link2">
    <material>Gazebo/Blue</material>
    <self_collide>1</self_collide>
  </gazebo>
  
  <gazebo reference="Link3">
    <material>Gazebo/SkyBlue</material>
    <self_collide>1</self_collide>
  </gazebo>
  
  <gazebo reference="Link4">
    <material>Gazebo/Purple</material>
    <self_collide>1</self_collide>
  </gazebo>
  
  <gazebo reference="Link5">
    <material>Gazebo/Orange</material>
    <self_collide>1</self_collide>
  </gazebo>
  
  <gazebo reference="Link6">
    <material>Gazebo/RedBright</material>
    <self_collide>1</self_collide>
  </gazebo>
  
  
  <!-- MOBILE ROBOT CONTROLLER (SKID STEER DRIVE PLUGIN - GAZEBO= -->
  <gazebo>
      <plugin name="skid_steer_drive_controller" filename="libgazebo_ros_skid_steer_drive.so">
          <always_on>1</always_on>
          <updateRate>100.0</updateRate>
          <robotNamespace>/</robotNamespace>
          <leftFrontJoint>joint_leftFrontWheel</leftFrontJoint>
          <rightFrontJoint>joint__rightFrontWheel</rightFrontJoint>
          <leftRearJoint>joint__leftRearWheel</leftRearJoint>
          <rightRearJoint>joint__rightRearWheel</rightRearJoint>
          <wheelSeparation>0.705</wheelSeparation>
          <wheelDiameter>0.100</wheelDiameter>
          <torque>1000</torque>
          <commandTopic>cmd_vel</commandTopic>
          <odometryTopic>odom</odometryTopic>
          <odometryFrame>odom</odometryFrame>
          <robotBaseFrame>dummy</robotBaseFrame>
          <publishWheelTF>true</publishWheelTF>
          <publishOdomTF>true</publishOdomTF>
          <publishWheelJointState>true</publishWheelJointState>
          <covariance_x>0.0001</covariance_x>
          <covariance_y>0.0001</covariance_y>
          <covariance_yaw>0.01</covariance_yaw>
          <broadcastTF>true</broadcastTF>
      </plugin>
  </gazebo>
  
  <!-- VIRTUAL IMU IN BASE_LINK (GAZEBO PLUGIN) -->
  <gazebo reference="base_link">
    <sensor type="imu" name="sensor_imu">
        <always_on>1</always_on>
        <update_rate>10</update_rate>
        <visualize>true</visualize>
        <plugin name="imu_plugin" filename="libgazebo_ros_imu_sensor.so">
          <topicName>/imu/data</topicName>
          <frameName>imu_frame</frameName>
          <updateRateHZ>100.0</updateRateHZ>
          <gaussianNoise>0.0</gaussianNoise>
          <xyzOffset>0 0 0</xyzOffset>
          <rpyOffset>0 0 0</rpyOffset>
      </plugin>
    </sensor>
 </gazebo>
 
 <!-- TRANSMISSION DEFINITION FOR EACH JOINT -->
 <transmission name="link_1_trans">
	<type>transmission_interface/SimpleTransmission</type>
	<joint name="Joint1">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	</joint>
	<actuator name="link_1_motor">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
		<mechanicalReduction>1</mechanicalReduction>
	</actuator>
  </transmission>
  <transmission name="link_2_trans">
	<type>transmission_interface/SimpleTransmission</type>
	<joint name="Joint2">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	</joint>
	<actuator name="link_2_motor">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
		<mechanicalReduction>1</mechanicalReduction>
	</actuator>
  </transmission>
  
  <transmission name="link_3_trans">
	<type>transmission_interface/SimpleTransmission</type>
	<joint name="Joint3">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	</joint>
	<actuator name="link_3_motor">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
		<mechanicalReduction>1</mechanicalReduction>
	</actuator>
  </transmission>
  
  <transmission name="link_4_trans">
	<type>transmission_interface/SimpleTransmission</type>
	<joint name="Joint4">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	</joint>
	<actuator name="link_4_motor">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
		<mechanicalReduction>1</mechanicalReduction>
	</actuator>
  </transmission>
  
  <transmission name="link_5_trans">
	<type>transmission_interface/SimpleTransmission</type>
	<joint name="Joint5">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	</joint>
	<actuator name="link_5_motor">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
		<mechanicalReduction>1</mechanicalReduction>
	</actuator>
  </transmission>
  
  <transmission name="link_6_trans">
	<type>transmission_interface/SimpleTransmission</type>
	<joint name="Joint6">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	</joint>
	<actuator name="link_6_motor">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
		<mechanicalReduction>1</mechanicalReduction>
	</actuator>
  </transmission>
  
  <!-- ARM ROBOT CONTROLLER (GAZEBO PLUGIN) -->
  <gazebo>
	<plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
		<robotNamespace>/</robotNamespace>
	</plugin>
  </gazebo>
  
  <!-- VIRTUAL IMU IN GLASS (GAZEBO PLUGIN) -->
  <gazebo reference="Link6">
    <sensor type="imu" name="sensor_imu2">
        <always_on>1</always_on>
        <update_rate>50</update_rate>
        <visualize>true</visualize>
        <plugin name="imu2_plugin" filename="libgazebo_ros_imu_sensor.so">
          <topicName>/imu2/data</topicName>
          <frameName>imu_frame2</frameName>
          <updateRateHZ>100.0</updateRateHZ>
          <gaussianNoise>0.0</gaussianNoise>
          <xyzOffset>0 0 0</xyzOffset>
          <rpyOffset>-3.14 0 0</rpyOffset>
          <broadcastTF>true</broadcastTF>
      </plugin>
    </sensor>
 </gazebo>

  
</robot>
